Fixed exceptional case handling in SherpaHadronizer

Dr15Jones opened the pull request:
#### PR description:

The MPI::Init function setups up a unix signal handler which interferes with the signal handled used in the cmsExternalGenerator program. Moving the call to the constructor allows the signal handler to be reset
before doing any data processing.

Even with this change, an unix signal happening in the constructor will cause a dealock of the ExternalGenerator system.

Also had to avoid a case where an exception could be thrown while we were unrolling from another exception. That would happen if doSetRandomEngine would throw an exception.

#### PR validation:

The code compiles and a simple job using the SherpaHadronizer with the ExternalGeneratorFilter ran fine.
Explicitly adding an exception to Sherpa allowed us to test the problem.

Dr15Jones committed:
Moved MPI::Init to constructor of SherpaHadronizer

The function setups up a unix signal handler which interferes with
the signal handled used in the cmsExternalGenerator program. Moving
the call to the constructor allows the signal handler to be reset
before doing any data processing.
Even with this change, an unix signal happening in the constructor
will cause a dealock of the ExternalGenerator system.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-35204/25144

- This PR adds an extra 16KB to repository

cmsbuild commented:
A new Pull Request was created by  @Dr15Jones (Chris Jones) for master.

It involves the following packages:

- GeneratorInterface/SherpaInterface (**generators**)


@SiewYan, @mkirsano, @Saptaparna, @cmsbuild, @GurpreetSinghChahal, @agrohsje, @alberto-sanchez can you please review it and eventually sign? Thanks.
@alberto-sanchez, @mkirsano, @thuer this is something you requested to watch as well.
@perrotta, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

Dr15Jones commented:
please test

Dr15Jones committed:
Protect RandomEngineSentry destructor from thrown an exception

SherpaHadronizer::doSetRandomEngine is called by the destructor
so throwing an exception from there can happen when we are already
dealing with an exception which leads to std::termiate.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-35204/25147

- This PR adds an extra 16KB to repository

cmsbuild commented:
Pull request #35204 was updated. @SiewYan, @mkirsano, @Saptaparna, @cmsbuild, @GurpreetSinghChahal, @agrohsje, @alberto-sanchez can you please check and sign again.

Dr15Jones commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d59747/18433/summary.html
**COMMIT**: ab7bcccb32aadec97821553c18e54918f67d444e
**CMSSW**: CMSSW_12_1_X_2021-09-08-1100/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/35204/18433/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 39
 - DQMHistoTests: Total histograms compared: 3001001
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3000973
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 38 files compared)
 - Checked 165 log files, 37 edm output root files, 39 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_1_X_2021-09-08-1100+d59747/45302/triggerResults): no differences found

Dr15Jones commented:
type bugfix

Dr15Jones commented:
urgent

we will need a backport to 10_6 for production.

GurpreetSinghChahal commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

perrotta commented:
+1

Dr15Jones commented on GeneratorInterface/SherpaInterface/src/SherpaHadronizer.cc:
```diff
@@ -210,14 +218,12 @@ SherpaHadronizer::~SherpaHadronizer() {
 
 bool SherpaHadronizer::initializeForInternalPartons() {
   //initialize Sherpa but only once
+  throw cms::Exception("TEST");
```
Argg! This was left over from my testing.

Dr15Jones replied:
Fixed in #35224
